### PR TITLE
Revert "Delete src directory"

### PR DESCRIPTION
**Pull request by TheRickyRoss2**
This reverts commit 292061a75b738d2adb908c5ff33fe1e05ec6e0be.

### CODE REVIEW

File: src/main/java/org/example/Twitter.java [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+import java.util.*;
+
+import java.util.Scanner;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.json.simple.parser.*;
+
+class Tweet {
```
Comment (TheRickyRoss2) [inline src/main/java/org/example/Twitter.java:13]: move unrelated classes to separate files

File: src/main/java/org/example/Twitter.java [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+    String content;
```
Comment (TheRickyRoss2) [inline src/main/java/org/example/Twitter.java:14]: make fields private

File: src/main/java/org/example/Twitter.java [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+        }
+    }
+
+    public static int login(String username, String password){
+        /*
+        Input: The username && password
+        Output: The username's userId or -1 if null and void
+        */
+
+        for(int i = 1; i < userDict.size()+1; i++) {
```
Comment (TheRickyRoss2) [inline src/main/java/org/example/Twitter.java:176]: reconsider a more efficient user login flow

File: src/main/java/org/example/Twitter.java [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+            String userId = Integer.toString(i);
+
+            String tempUser = userDict.get(userId).username;
```
Comment (TheRickyRoss2) [inline src/main/java/org/example/Twitter.java:179]: do not user direct member access, use getters

File: src/main/java/org/example/Twitter.java [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+        for(int i = 1; i < tweetDict.size()+1; i++) {
+            ArrayList<String> temp = new ArrayList<>();
+            String tweetId = Integer.toString(i);
+            String userId = Integer.toString(tweetDict.get(tweetId).userId);
+
+            System.out.print("Tweet Id: " + tweetId);
+            System.out.print(" / User: " + userDict.get(userId).name);
+            System.out.println(" / Content: " + tweetDict.get(tweetId).content);
+
+            Collections.addAll(temp, tweetId, userDict.get(userId).name, tweetDict.get(tweetId).content);
```
Comment (TheRickyRoss2) [inline src/main/java/org/example/Twitter.java:289]: consider response object to hold the information in a structured manner

File: src/main/java/org/example/Twitter.java [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+                makeTweet(textBody, hashtags);
+
+            } else {
+                s = "quit";
+            }
+        }
+        in.close();
+    }
+
+    public static ArrayList<ArrayList<String>> showAllTweets(){
```
Comment (TheRickyRoss2) [inline src/main/java/org/example/Twitter.java:271]: do you really need to return anything here? consider what happens with a large dataset wrt memory consumption

File: src/main/java/org/example/Twitter.java [hunk 7]
```diff
@@ -0,0 +1,461 @@
...
+
+} 
+
+
+public class Twitter
+{
+    public static int currentUserId = -1;
+    public static int tweetCounter = 0;
+    public static LinkedHashMap<String, Tweet> tweetDict = new LinkedHashMap<String, Tweet>();
+    public static LinkedHashMap<String, User> userDict = new LinkedHashMap<String, User>();
```
Comment (TheRickyRoss2) [inline src/main/java/org/example/Twitter.java:136]: can be compressed to list for memory efficiency given unique incremental integer identifiers

File: src/main/java/org/example/Twitter.java [hunk 8]
```diff
@@ -0,0 +1,461 @@
...
+         return results;
+     }
+
+     public static ArrayList<Object> makeTweet(String userText, ArrayList<String> userHashtags){
+        /*
+        Input: User's text && user's hashtags
+        Output: An array containing both the user's text and user's hashtags
+
+        Function adds in the new tweet into the current json file.
+        */
```
Comment (TheRickyRoss2) [inline src/main/java/org/example/Twitter.java:395]: nit: use javadocs format

File: src/main/java/org/example/Twitter.java [hunk 9]
```diff
@@ -0,0 +1,461 @@
...
+         ArrayList<Object> results = new ArrayList<>();
+         Collections.addAll(results, userText, userHashtags);
+
+         //Write to tweetDict
+         System.out.println("// Write New Tweet //");
+
+         tweetCounter++;
+         ArrayList<String> newHashtags = userHashtags;
+
+         Tweet newTweet = new Tweet();
```
Comment (TheRickyRoss2) [inline src/main/java/org/example/Twitter.java:406]: consider builder class

**Comment by TheRickyRoss2** [comment]
describe how you would support N concurrent users

**Comment by TheRickyRoss2** [comment]
describe what would change if used rdb

**Comment by TheRickyRoss2** [comment]
describe how to scale search

**Comment by TheRickyRoss2** [comment]
describe implementation limitations and known inefficiencies (i.e. support N tweets/seconds, program takes M seconds to startup, only supports X users, Y tweets)

**Comment by Azirly** [comment]
- **describe implementation limitations and known inefficiencies (i.e. support N tweets/seconds, program takes M seconds to startup, only supports X users, Y tweets)**
    * Limitations
        - Can only support 1 user using the command line at a time
        - Can support around 14,285,714 tweets/users as the script is limited by the database (i.e. 1 GB)
        - Due to the size of the script, it takes less than 1 second to bootup
    * Other Limitations:
        - Scalability
            * Using a single JSON file is not scalable, as it will lead slow performance once enough users/tweets are added in
        - Security
            * Using a single JSON file lacks the security as unauthorized individuals can go and access the file
        - User Experience/Features
            * The command line interface is worse for a social media implementation when compared to a general GUI, as the CLI makes tasks like uploading pictures and photos much harder
        - Data Backup
            * A single json file means that if the json file is deleted/corrupted, then the data is lost
        - Limited Search Functionality
            * Searching for tweets in a single json file will be limited once enough users/tweeets are on the file
        - Limited Analytics
            * The current iteration of the program does not contain any user tracking, making it difficult to retrieve information to improve on the application
        - Limited Integration
            * Due to the CLI design, there is no possible integration with other 3rd-party services
- **describe how to scale search**
    * Database Upgrading:
        - MySQL/MongoDB
        - For more efficient searching && scalability
        - For storing large # of files and documents
    * Search Engine
        - Elasticsearch/Solr
        - Allows for full-text search, relevance ranking, faceted search
    * Caching
        - Store frequently searched tweets in memory
        - Reduce # of times needed to access the json file
    * Multithreaading
        - Allow for faster search times && better scalability
    * Cloud-based storage
        - AWS S3/Google Cloud Storage/Azure Blob Storage
        - For storing structured data that needs to be analyzed
    * Search Algorithm:
        - Like a trie (aka phonebook)
    * Load Balancer
        - distribute the request among multiple servers so the app is responsive even under heavy loads
- **describe what would change if used rdb**
    * TL;DR:
        - Using a redis database would solve most of the backend limitations with my project
    * In-memory storage:
        - Redis stores data in-memory, allowing for faster reads and write operations
    * Caching:
        - Redis uses a caching layer where I can store frequently accessed tweets in memory and improve search performance
    * Pub/Sub:
        - Redis uses a publish/subscribe mechanism, allowing me to send real-time updates to users && improve the user experience
    * Scaling:
        - Redis uses a master/slave replication && sharding, allowing for horizontal scaling && large data handles
    * Data Backup:
        - Redis supports a built-in persistence mechanism, aka handling any data loss
    * Searching:
        - Redis supports full-text search, allowing for more optimized searching like fuzzy search, phrase search, and proximity search
    * Performance:
        - Redis is known for high peformance && low latency, allowing one to handle a large # of requests && fast response times to users
- **describe how you would support N concurrent users**
    * In addition to using redis to handle load balancing, caching, multi-threading, here are other strategies to support N concurrent users
    * Cloud-based infrastructure:
        - AWS/Azure/Google
    * Session Management:
        - to handle user sessions && keep track of user information
    * Asynchronous Processing:
        - Handle user requests asynchronously so that we can deal with them simultaneously
    * Monitoring and Logging
    * Distributed Systems:
        - Like using a message queue to handle requests.

File: src/main/java/org/example/Twitter.java (hunk 7 above)
Comment (Azirly) [reply src/main/java/org/example/Twitter.java:136]: Using usernames as identifiers, this code needs a hashmap.

File: src/main/java/org/example/Twitter.java (hunk 6 above)
Comment (Azirly) [reply src/main/java/org/example/Twitter.java:271]: The idea was to have output for jUnit testing. So it depends if we care for a more stable codebase or less memory consumption.

File: src/main/java/org/example/Twitter.java (hunk 3 above)
Comment (Azirly) [reply src/main/java/org/example/Twitter.java:176]: User Login Flow is now O(1) rather than O(n)

**Review by Azirly — COMMENTED** [review]
All other comments you have made I have applied to my project
